src/Peers/provider.cpp

```diff
@@ -59,80 +61,94 @@ void Provider::listen() {
-            Message aggregateResultMsg(uuid, IpAddress(host, port),
-                                       aggregatePayload);
+        // append to followerData
+        followerData.push_back(taskResp->getTrainingData());
+        server->replyToConn("Received follower result.");
+        server->closeConn();
+    }
 
-            cout << "Waiting for connection back to requester" << endl;
-            // busy wait until connection is established
-            while (client->setupConn(reqHost, reqPort, "tcp") == -1) {
-                sleep(5);
-            }
+    cout << endl;
+    workloadThread.join();
+    TaskResponse aggregatedResults = aggregateResults(followerData);
 
-            client->sendRequest(aggregateResultMsg.serialize().c_str());
-        } else {
-            // if this is a follower, send the result to the leader peer
-            cout << "Waiting for connection back to leader" << endl;
-            // get the ip from the bootstrap server
-            // ------------------ hard code for now ------------------
-            IpAddress leaderIp =
-                task->getAssignedWorkers()[task->getLeaderUuid()];
-            const char* leaderHost = leaderIp.host.c_str();
-            const char* leaderPort = to_string(leaderIp.port).c_str();
-            // busy wait until connection is established with the leader
-            while (client->setupConn(leaderHost, leaderPort, "tcp") == -1) {
-                sleep(5);
+    // Send results back to requester
+    // TODO: requester IP address could change
+    shared_ptr<TaskResponse> aggregatePayload =
+        make_shared<TaskResponse>(aggregatedResults);
+    Message aggregateResultMsg(uuid, IpAddress(host, port),
+                                aggregatePayload);
+
+    cout << "Waiting for connection back to requester" << endl;
+    // Keep trying to send results back to requester 
+    while (true) {
+        // busy wait until connection is established
+        while (client->setupConn(requesterIpAddr, "tcp") != 0) {
+            sleep(5);
+        }
+        if (client->sendMsg(aggregateResultMsg.serialize().c_str()) != 0) {
+            continue;
+        }
+
+        if (!server->acceptConn()) continue;
+
+        // receive response from requester
+        string serializedData = server->receiveFromConn();
+        server->closeConn();
+        // process this request
+        Message msg;
+        msg.deserialize(serializedData);
+        shared_ptr<Payload> payload = msg.getPayload();
+
+        if (payload->getType() == Payload::Type::GENERIC_RESPONSE) {
+            shared_ptr<GenericResponse> gr =
+                static_pointer_cast<GenericResponse>(payload);
+            if (gr->getStatus() == GenericResponse::Status::SUCCESS) {
+                break;
             }
-            workloadThread.join();
-            client->sendRequest(task->serialize().c_str());
         }
+    }
+}
 
+void Provider::followerHandleTaskRequest() {
+    processWorkload();
```

Comment:
Is there a reason why the follower wouldn't execute this in a thread, similar to leader?

Reply:
The leader starts another thread because it needs to do the training work and collect results from the follower peers at the same time. However, this code will soon be changed once we implement our multi-process implementation separating training and p2p into different processes.